ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -15,52 +15,256 @@
+      newMetric.add(metric);
+      // create gauge for task level storage usage
+      metricRegistry.addMetric(newMetric.metricName, (Gauge<BigInteger>) (config, now) -> newMetric.getValue());
+
+      // add to list of seen task metrics for this query
+      metricsSeen.get(queryId).add(newMetric);
+    }
   }
+
+  @Override
+  public void metricRemoval(final KafkaMetric metric) {
+    final MetricName metricName = metric.metricName();
+    if (!metricName.name().equals("total-sst-files-size")) {
+      return;
+    }
+
+    final String queryId = metric.metricName().tags().getOrDefault("app-id", "");
+    final List<TaskStorageMetric> taskMetrics = metricsSeen.get(queryId);
+    // will delete task / query gauges if we don't have any more metrics for them
+    for (TaskStorageMetric m : taskMetrics) {
+      // if don't have any store metrics for this task
+      m.remove(metric);
+      if (m.metrics.size() == 0) {
+        // we can get rid of task level metric, there are no sst-metrics anymore
+        metricRegistry.removeMetric(m.metricName);
+        metricsSeen.get(queryId).remove(m);
+        // if there are no task level metrics for this query
+        if (taskMetrics.size() == 0){
+          // we don't have more task metrics for the query, remove query gauge
+          metricRegistry.removeMetric(metricRegistry.metricName(
+            "query-storage-usage",
+            "ksql-utilization-metrics",
+            ImmutableMap.of("query-id", queryId))
+          );
+          metricsSeen.remove(queryId);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public Set<String> reconfigurableConfigs() {
+    return null;
+  }
+
+  @Override
+  public void validateReconfiguration(Map<String, ?> configs) throws ConfigException {
+
+  }
+
+  @Override
+  public void reconfigure(Map<String, ?> configs) {
+
+  }
+
+  @Override
+  public void contextChange(MetricsContext metricsContext) {
+
+  }
+
+  public void resetAndUpdateMetric(final KafkaMetric metric, final String taskId, final String queryId) {
+    final List<TaskStorageMetric> taskMetrics = metricsSeen.get(queryId);
+    taskMetrics.forEach(r -> r.remove(metric));
+    for (TaskStorageMetric storageMetric : taskMetrics) {
+      if (storageMetric.getTaskId().equals(taskId)) {
+        storageMetric.add(metric);
+      }
+    }
+  }
+
+  final BigInteger computeQueryMetric(final String queryId) {
+    BigInteger queryMetricSum = BigInteger.ZERO;
+    for (TaskStorageMetric m : metricsSeen.get(queryId)) {
+      queryMetricSum = queryMetricSum.add(m.getValue());
+    }
+    return queryMetricSum;
+  }
+
+  public static class TaskStorageMetric {
+    final MetricName metricName;
+    private final Map<MetricName, KafkaMetric> metrics = new ConcurrentHashMap<>();
+
+    TaskStorageMetric(final MetricName metricName) {
+      this.metricName = metricName;
+    }
+
+    private void add(final KafkaMetric metric) {
```

Comment:
add, remove, and getValue need to be synchronized - streams threads may be creating/removing state stores while telemetry is reading the value

Reply:
never mind - noticed this is a `ConcurrentHashMap`